src/Kestrel.Core/Internal/BufferReader.cs

```diff
@@ -1,153 +1,494 @@
+
+            GetFirstSpan(buffer, out ReadOnlySpan<T> first, out _nextPosition);
+            CurrentSpan = first;
+            _moreData = first.Length > 0;
+
+            if (!buffer.IsSingleSegment && !_moreData)
+            {
+                _moreData = true;
+                GetNextSpan();
+            }
+        }
 
-        private int _consumedBytes;
-        private bool _end;
+        private const int FlagBitMask = 1 << 31;
+        private const int IndexBitMask = ~FlagBitMask;
 
+        // TODO:
+        // Move to helper in ReadOnlySequence
+        // https://github.com/dotnet/corefx/issues/33029
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public BufferReader(in ReadOnlySequence<byte> buffer)
+        private static void GetFirstSpan(in ReadOnlySequence<T> buffer, out ReadOnlySpan<T> first, out SequencePosition next)
         {
-            _index = 0;
-            _consumedBytes = 0;
-            _sequence = buffer;
-            _currentSequencePosition = _sequence.Start;
-            _nextSequencePosition = _currentSequencePosition;
+            first = default;
+            next = default;
+            SequencePosition start = buffer.Start;
+            int startIndex = start.GetInteger();
+            object startObject = start.GetObject();
 
-            if (_sequence.TryGet(ref _nextSequencePosition, out var memory, true))
+            if (startObject != null)
             {
-                _end = false;
-                _currentSpan = memory.Span;
-                if (_currentSpan.Length == 0)
+                SequencePosition end = buffer.End;
+                int endIndex = end.GetInteger();
+                bool isMultiSegment = startObject != end.GetObject();
+
+                // A == 0 && B == 0 means SequenceType.MultiSegment
+                if (startIndex >= 0)
                 {
-                    // No space in first span, move to one with space
-                    MoveNext();
+                    if (endIndex >= 0)  // SequenceType.MultiSegment
+                    {
+                        ReadOnlySequenceSegment<T> segment = (ReadOnlySequenceSegment<T>)startObject;
+                        next = new SequencePosition(segment.Next, 0);
+                        first = segment.Memory.Span;
+                        if (isMultiSegment)
+                        {
+                            first = first.Slice(startIndex);
+                        }
+                        else
+                        {
+                            first = first.Slice(startIndex, endIndex - startIndex);
+                        }
+                    }
+                    else
+                    {
+                        if (isMultiSegment)
+                            ThrowInvalidOperationException_EndPositionNotReached();
+
+                        first = new ReadOnlySpan<T>((T[])startObject, startIndex, (endIndex & IndexBitMask) - startIndex);
```

Comment:
How do we know that startObject is an array?

Reply:
startIndex is positive, endIndex is negative which corresponds to array sequence type https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L551-L552